### PR TITLE
rollback: use 237 wal pkg

### PR DESCRIPTION
**Pull request by xiang90**
/cc @wojtek-t @mml

### CODE REVIEW

File: cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2015 CoreOS, Inc.
```
Comment (wojtek-t) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: I think it will not pass our verify scripts, since it requires kubernetes-authors copyright.

I'm not sure how to solve this problem though - ideally it should go to vendor/ dir, but I'm not sure if this is possible.

Comment (xiang90) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: hmm... this modified wal pkg only lives here now. I did a few modification to keep the dependencies minimal. I am not sure how to solve it. We do not want to have this thing in etcd repo.

Comment (wojtek-t) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: I agree that it shouldn't live in etcd repo.

@mml - can you please talk with people from MTV to check how this can be solved? Maybe @thockin has some ideas?

Comment (mikedanese) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: you can put it in third_party/forked/

Comment (xiang90) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: @mml @wojtek-t 

Can any of you verify the current solution can work and build?

If everyone is happy about moving the modified dep to third_part/forked/, I will do this after we feel good about this solution.

Comment (thockin) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: Is this code that is now part of Kubernetes (i.e. being donated by the authors/copyright holders) or code that we're forking and tweaking?

The Copyright line is pretty much legally meaningless, but it's required nonetheless.

Comment (xiang90) [inline cluster/images/etcd/rollback/etcd237/pkg/fileutil/fileutil.go]: @wojtek-t How shall we move this forward? Has mml verified this? Can you check again with him? Thank you!

**Comment by mml** [comment]
I was out a lot last week.  I'm copying your branch right now.

**Comment by mml** [comment]
This won't build for me.  @xiang90 were you able to build it?

```
docker run -it -v /usr/local/google/home/mml/gocode/src/k8s.io/kubernetes/cluster/images/etcd/../../../:/go/src/k8s.io/kubernetes -v /usr/local/google/home/mml/gocode/src/k8s.io/kubernetes/cluster/images/etcd/tmp:/build -e GOARCH=amd64 golang:1.6.3 \
                /bin/bash -c "CGO_ENABLED=0 go build -o /build/rollback k8s.io/kubernetes/cluster/images/etcd/rollback"
src/k8s.io/kubernetes/cluster/images/etcd/rollback/etcd237/wal/util.go:22:2: cannot find package "github.com/coreos/etcd237/pkg/fileutil" in any of:
        /go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd237/pkg/fileutil (vendor tree)
        /usr/local/go/src/github.com/coreos/etcd237/pkg/fileutil (from $GOROOT)
        /go/src/github.com/coreos/etcd237/pkg/fileutil (from $GOPATH)
make: *** [build] Error 1
```

**Comment by xiang90** [comment]
@mml Can you try again with my last push? I was not on a clean GOPATH. Now I can build this in the rollback folder by doing go build.

**Comment by mml** [comment]
@xiang90 that builds.  Running it now...

**Comment by mml** [comment]
OK.  It runs to completion, but when trying to bring up 2.3.7 with the rolled back data, I get the following message in the etcd log and etcd does not come up: `C | etcdserver: cluster cannot be downgraded (current version: 2.3.7 is lower than determined cluster version: 2.3).`

Here is the entire log, and I can upload a tarball of the post-rollback data if needed.

```
 === /usr/local/bin/etcd --debug --name etcd-master --listen-peer-urls http://127.0.0.1:2380 --initial-advertise-peer-urls http://127.0.0.1:2380 --advertise-client-urls http://127.0.0.1:2379 --listen -client-urls http://127.0.0.1:2379 --data-dir /var/etcd/data --initial-cluster-state new --initial-cluster etcd-master=http://127.0.0.1:2380 
 2016-12-12 22:46:58.081714 I | etcdmain: etcd Version: 2.3.7
 2016-12-12 22:46:58.081793 I | etcdmain: Git SHA: fd17c91
 2016-12-12 22:46:58.081806 I | etcdmain: Go Version: go1.6.2
 2016-12-12 22:46:58.081819 I | etcdmain: Go OS/Arch: linux/amd64
 2016-12-12 22:46:58.081832 I | etcdmain: setting maximum number of CPUs to 12, total number of available CPUs is 12
 2016-12-12 22:46:58.081890 W | etcdmain: found invalid file/dir version.txt under data dir /var/etcd/data (Ignore this if you are upgrading etcd)
 2016-12-12 22:46:58.081906 N | etcdmain: the server is already initialized as member before, starting as etcd member...
 2016-12-12 22:46:58.081999 I | etcdmain: listening for peers on http://127.0.0.1:2380
 2016-12-12 22:46:58.082063 I | etcdmain: listening for client requests on http://127.0.0.1:2379
 2016-12-12 22:46:58.090387 I | etcdserver: recovered store from snapshot at index 62051
 2016-12-12 22:46:58.090412 I | etcdserver: name = etcd-master
 2016-12-12 22:46:58.090419 I | etcdserver: data dir = /var/etcd/data
 2016-12-12 22:46:58.090426 I | etcdserver: member dir = /var/etcd/data/member
 2016-12-12 22:46:58.090433 I | etcdserver: heartbeat = 100ms
 2016-12-12 22:46:58.090439 I | etcdserver: election = 1000ms
 2016-12-12 22:46:58.090444 I | etcdserver: snapshot count = 10000
 2016-12-12 22:46:58.090455 I | etcdserver: advertise client URLs = http://127.0.0.1:2379
 2016-12-12 22:46:58.091023 I | etcdserver: restarting member 58f5376e190001 in cluster 58f5376e190002 at commit index 0
 2016-12-12 22:46:58.091282 I | raft: 58f5376e190001 became follower at term 0
 2016-12-12 22:46:58.091360 I | raft: newRaft 58f5376e190001 [peers: [58f5376e190001], term: 0, commit: 62051, applied: 62051, lastindex: 62051, lastterm: 6]
 2016-12-12 22:46:58.092134 C | etcdserver: cluster cannot be downgraded (current version: 2.3.7 is lower than determined cluster version: 2.3).
```

**Comment by xiang90** [comment]
@mml   I believe the problem occurs since the build is not exactly the same as the binary I gave you.  Can you share me with the tar? I can fix this and verify shortly.

**Comment by mml** [comment]
The latest version of the code works with my test cluster.  @thockin how do we move the licensing stuff forward?

**Comment by wojtek-t** [comment]
@xiang90 @mml - I think that moving this to third_party (as @mikedanese suggested) is the way to proceed.

**Comment by xiang90** [comment]
@wojtek-t @mml OK. Fixed. PTAL.

File: cluster/images/etcd/rollback/rollback.go [hunk 2]
```diff
@@ -42,7 +44,7 @@ import (
 	"github.com/golang/glog"
 )
 
-const rollbackVersion = "2.3.7"
+const rollbackVersion = "2.2.0"
```
Comment (wojtek-t) [inline cluster/images/etcd/rollback/rollback.go]: Why this change?

File: cluster/images/etcd/rollback/rollback.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ import (
 	"strings"
```
Comment (wojtek-t) [inline cluster/images/etcd/rollback/rollback.go]: Please fix verify-bazel.

File: cluster/images/etcd/rollback/rollback.go (hunk 2 above)
Comment (wojtek-t) [inline cluster/images/etcd/rollback/rollback.go]: I don't understand. This too is supposed to rollback from 3.0.* format to 2.3.7. Why do we need any information about 2.2.0?

File: cluster/images/etcd/rollback/rollback.go (hunk 3 above)
Comment (xiang90) [inline cluster/images/etcd/rollback/rollback.go]: What is verify-bazel?

Comment (mml) [inline cluster/images/etcd/rollback/rollback.go]: Run `hack/verify-bazel.sh`

**Comment by mml** [comment]
Also, can you leave a comment in `third_party/forked/etcd237` that explains why this is here and which sha1 it originates from?

**Comment by mml** [comment]
The comment should go in something like `README.md`.

**Comment by xiang90** [comment]
@mml Sure. I will address your comments on third_party thing tomorrow.

File: cluster/images/etcd/rollback/rollback.go (hunk 3 above)
Comment (mikedanese) [inline cluster/images/etcd/rollback/rollback.go]:
```
W1215 09:36:03.180] wrote BUILD for "/go/src/k8s.io/kubernetes/third_party/forked/etcd237/pkg/fileutil"
W1215 09:36:03.195] wrote BUILD for "/go/src/k8s.io/kubernetes/third_party/forked/etcd237/wal"
W1215 09:36:03.196] extract err: cannot find package "github.com/coreos/etcd237/Godeps/_workspace/src/github.com/gogo/protobuf/proto" in any of:
W1215 09:36:03.196] 	/go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd237/Godeps/_workspace/src/github.com/gogo/protobuf/proto (vendor tree)
W1215 09:36:03.196] 	/usr/local/go/src/github.com/coreos/etcd237/Godeps/_workspace/src/github.com/gogo/protobuf/proto (from $GOROOT)
W1215 09:36:03.196] 	/go/src/github.com/coreos/etcd237/Godeps/_workspace/src/github.com/gogo/protobuf/proto (from $GOPATH)
W1215 09:36:03.197] wrote BUILD for "/go/src/k8s.io/kubernetes/third_party/forked/etcd237/wal/walpb"
```

This says two things.
1. this package won't compile because [this import](https://github.com/kubernetes/kubernetes/blob/d26bf3195007534453c2324ef18e95694b5a9a46/third_party/forked/etcd237/wal/walpb/record.pb.go#L20) is not in our gopath. I can verify this by checking out your branch and running `go install ./third_party/forked/etcd237/...`. I get a compilation error.
2. `BUILD` files need to be update in `third_party/forked/etcd237/wal`, `third_party/forked/etcd237/pkg/fileutil`, and `third_party/forked/etcd237/wal/walpb` which makes sense because these are brand new packges. You can do this by rebasing this PR onto latest master and running `./hack/update-bazel.sh` then commiting the changes.

Comment (xiang90) [inline cluster/images/etcd/rollback/rollback.go]: @mikedanese 

OK. Now I see them... The default output from build-log.txt is very confusing since it hides all valuable output with `... skipping 1747 lines ...` and printed out tons of unrelated errors with yellow highlights.

```
W1215 09:44:40.288] I1215 09:44:40.287939   10387 parse.go:353] type checker error: /usr/local/go/src/runtime/trace.go:139:30: cannot use (traceBufHeader literal) (value of type traceBufHeader) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
W1215 09:44:40.289] I1215 09:44:40.288070   10387 parse.go:353] type checker error: /usr/local/go/src/runtime/trace.go:139:7: array length 64 << 10 - unsafe.Sizeof((traceBufHeader literal)) (value of type uintptr) must be constant
W1215 09:44:40.289] I1215 09:44:40.288156   10387 parse.go:353] type checker error: /usr/local/go/src/runtime/mgcwork.go:285:37: cannot use (workbufhdr literal) (value of type workbufhdr) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
W1215 09:44:40.289] I1215 09:44:40.288181   10387 parse.go:353] type checker error: /usr/local/go/src/runtime/mgcwork.go:285:7: array length (_WorkbufSize - unsafe.Sizeof((workbufhdr literal))) / sys.PtrSize (value of type uintptr) must be constant
W1215 09:44:40.289] I1215 09:44:40.288556   10387 parse.go:353] type checker error: /usr/local/go/src/runtime/chan.go:21:28: cannot use (hchan literal) (value of type hchan) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
W1215 09:44:40.289] I1215 09:44:40.288590   10387 parse.go:353] type che
```

We probably need to improve this output format to make it less confusing.

I will play with bazel more tomorrow.

File: cluster/images/etcd/rollback/rollback.go (hunk 2 above)
Comment (wojtek-t) [inline cluster/images/etcd/rollback/rollback.go]: Got it - thanks for explanation.

**Comment by xiang90** [comment]
@mml  all fixed.

**Comment by wojtek-t** [comment]
@xiang90 - thanks!

/lgtm
